### PR TITLE
chore: Rewrite rbac rego -> SQL clause

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

The previous code became hard to follow as we added more edge cases. All these edge cases happened because of our jsonb columns in SQL. This new code removes all those edge cases from the general code, as has the exceptions passed into the package. This code handles our custom jsonb columns:

https://github.com/coder/coder/blob/stevenmasley/rego_to_sql_redo/coderd/rbac/regosql/aclGroupVar.go

The previous code had a lot of stuff like this that was hard to follow. Changing it would cause random side effects, something we don't want in our auth code.

- https://github.com/coder/coder/blob/main/coderd/rbac/query.go#L288-L316
- https://github.com/coder/coder/blob/main/coderd/rbac/query.go#L457-L476
- Werid type: https://github.com/coder/coder/blob/main/coderd/rbac/query.go#L17

# Implementation Notes

## GetAuthorizedTemplates

I added `GetAuthorizedTemplates` and added a great unit test which tests all this SQL.

## Testing

Testing is critical as this package generates SQL clauses, but does not actually execute the SQL clauses. It is possible the SQL that is generated is either syntactically incorrect for SQL, or behaviorally incorrect. It is up to us to come up with the SQL we desire, and make unit tests with these asserted cases.

## Implementing Types

Types implement their supported operations on the [types themselves](https://github.com/coder/coder/blob/300dc4ccb65dee8fe8fb4b835e30654c06f9f21a/coderd/rbac/regosql/sqltypes/string.go#L20). This is opposed to implementing a nested conditional structure such as:

```golang
// Trying to do a = b
switch a.(type) {
  case string:
    switch b.(type): 
      case string:
        return fmt.Sprintf("%s = %s, a, b)
}
```

This nesting structure blows up quick and becomes very difficult to follow. Implementing it on the type is much more straight forward and easier to understand. It also allows for [easy recursion](https://github.com/coder/coder/blob/300dc4ccb65dee8fe8fb4b835e30654c06f9f21a/coderd/rbac/regosql/sqltypes/equality.go#L63) for things like `(a = b) = (c = d)`. This is 2 equalities being compared, and in the structure above things like this become quite a hassle.

The other benefit to this, is that we can override behavior outside the package. The biggest problem in the first implementation is that there is no easy way to handle arbitrary jsonb types. I made an attempt at this by creating a `Tree` type for variables, but it's unnecessarily complex for the 1 column type we care about that is the exception. In this new design the exception is a [custom type](https://github.com/coder/coder/blob/300dc4ccb65dee8fe8fb4b835e30654c06f9f21a/coderd/rbac/regosql/aclGroupVar.go) passed into the generator. So the generation code does not need to handle these [like it did before](https://github.com/coder/coder/blob/015a6f9e2665ddc08687bd0a304a3f05439c2ffc/coderd/rbac/query.go#L289-L301).

### CODE REVIEW

**Review by kylecarbs — APPROVED** [review]
Approving this so the improvement can get in. Did you run the benchmarks against this change?

**Review by mafredri — APPROVED** [review]
Looks great, there's quite a few bits I don't fully understand due to being unfamiliar with rego, but it looks solid. It'd be nice to find an alternative solution for re-writing the generated DB query functions, but not sure what that would look like.

Maybe sqlc plugins could be utilized there? 🤔 https://docs.sqlc.dev/en/latest/guides/plugins.html

File: coderd/rbac/regosql/sqltypes/or.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package sqltypes
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/rbac/regosql/sqltypes/or.go:1]: ah yes. It was originally just `OR` :+1:

File: coderd/rbac/regosql/compile.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		return tyBool, nil
+	}
+
+	return nil, xerrors.Errorf("expression %s not supported", e.String())
+}
+
+// convertCall converts a function call to a SQL expression.
+func convertCall(cfg ConvertConfig, call ast.Call) (sqltypes.Node, error) {
+	// Operator is the first term
+	op := call[0]
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/rbac/regosql/compile.go:114]: Yes. But I can put a defensive check in case it is not used correctly.

The length 3 (really 2 after the op) is checked in `convertTerms(_, _ 2)`. So I know there is 2 args in the slice.

**Comment by Emyrk** [comment]
> Looks great, there's quite a few bits I don't fully understand due to being unfamiliar with rego, but it looks solid. It'd be nice to find an alternative solution for re-writing the generated DB query functions, but not sure what that would look like.
> 
> Maybe sqlc plugins could be utilized there? thinking https://docs.sqlc.dev/en/latest/guides/plugins.html

I have not looked into their plugins. That would be awesome if we could extend sqlc and not write custom db code.

**Comment by Emyrk** [comment]
@kylecarbs just pinging about what @mafredri linked: https://docs.sqlc.dev/en/latest/guides/plugins.html. Would be cool to use that instead of `modelqueries.go`

**Review by f0ssel — APPROVED** [review]
I did not review the code in the `rego` directory since I was very unfamiliar with it, but everything else regarding `coderd` seems reasonable. Agree about extending sqlc, we know that we can with plugins but because you need to integrate with a standalone binary it's more work that just some code. I'd be very happy if we did the upfront work to make a plugin if it resulted in our `modelqueries.go` workaround going away.
